### ISSUE TITLE
4.0.3: TS2322: Type 'HtmlWebpackPlugin' is not assignable to type 'Plugin'

### ISSUE THREAD

**Issue by stilettk** [jantimon/html-webpack-plugin#1383]
## Expected behaviour

Build should succeed like in v4.0.2.

## Current behaviour

```
TSError: ⨯ Unable to compile TypeScript:
webpack.config.ts:59:3 - error TS2322: Type 'HtmlWebpackPlugin' is not assignable to type 'Plugin'.
  Types of property 'apply' are incompatible.
    Type '(compiler: import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/html-webpack-plugin/node_modules/@types/webpack/index").Compiler) => void' is not assignable to type '(compiler: import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/@types/webpack/index").Compiler) => void'.
      Types of parameters 'compiler' and 'compiler' are incompatible.
        Type 'import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/@types/webpack/index").Compiler' is not assignable to type 'import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/html-webpack-plugin/node_modules/@types/webpack/index").Compiler'.
          The types of 'hooks.shouldEmit.call' are incompatible between these types.
            Type '(arg1?: import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/@types/webpack/index").compilation.Compilation | undefined, arg2?: any, arg3?: any, ...args: any[]) => any' is not assignable to type '(arg1?: import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/html-webpack-plugin/node_modules/@types/webpack/index").compilation.Compilation | undefined, arg2?: any, arg3?: any, ...args: any[]) => any'.
              Types of parameters 'arg1' and 'arg1' are incompatible.
                Type 'import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/html-webpack-plugin/node_modules/@types/webpack/index").compilation.Compilation | undefined' is not assignable to type 'import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/@types/webpack/index").compilation.Compilation | undefined'.                  Type 'import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/html-webpack-plugin/node_modules/@types/webpack/index").compilation.Compilation' is not assignable to type 'import("C:/Users/n.batov/RiderProjects/restaurantcashier-frontend/node_modules/@types/webpack/index").compilation.Compilation'.

 59     new HtmlWebpackPlugin({
        ~~~~~~~~~~~~~~~~~~~~~~~
 60             filename: 'index.html',
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
...
 63             favicon: resolve(__dirname, 'public/content/images/favicon.ico'),
    ~~~~~~~~~
```

## Environment

Tell us which operating system you are using, as well as which versions of Node.js, npm, webpack, and html-webpack-plugin. Run the following to get it quickly:

```
node -e "var os=require('os');console.log('Node.js ' + process.version + '\n' + os.platform() + ' ' + os.release())"
Node.js v12.2.0
win32 10.0.18363

npm --version
6.9.0

npm ls webpack
`-- webpack@4.42.1

npm ls html-webpack-plugin
`-- html-webpack-plugin@4.0.3
```

### Config

Copy the minimal `webpack.config.js` to produce this issue:

I use typescript in `webpack config.ts`

```ts
import HtmlWebpackPlugin from 'html-webpack-plugin'
...
plugins: [
...
	module: {
		rules: [
			{
				options: {
					transpileOnly: true,
					configFile: 'tsconfig.build.json',
				},
				test: /\.(js|tsx?)$/,
				exclude: /node_modules/,
				loader: 'ts-loader',
			},
			{ test: /\.css$/, use: ['style-loader'] },
			{ test: /\.(png|jpe?g|gif|woff(2)?)$/, loader: 'file-loader' },
			{
				test: /\.svg$/,
				use: ['@svgr/webpack', 'file-loader'],
			},
		],
	},
...
	new HtmlWebpackPlugin({
		filename: 'index.html',
		chunks: ['main'],
		template: resolve(__dirname, 'public/index.html'),
		favicon: resolve(__dirname, 'public/content/images/favicon.ico'),
	}),
...
]
```

## Additional context

**The problem is not reproduced in version 4.0.2**

**Comment by argv-minus-one** [comment]
The problem is that there is both a `node_modules/@types/webpack` and a `node_modules/html-webpack-plugin/node_modules/@types/webpack` with incompatible type definitions.

This happens because `html-webpack-plugin` has a dependency on `@types/webpack`, and so does your project, but your project depends on an *older* version of `@types/webpack`. Npm, silly thing that it is, handles this situation by silently installing a separate, incompatible copy of `@types/webpack` just for `html-webpack-plugin`!

You can fix it by running `npm update @types/webpack`, then `npm dedupe`. This should remove `node_modules/html-webpack-plugin/node_modules/@types/webpack` and make TypeScript happy.

---

To prevent this kind of version conflict, `html-webpack-plugin` could declare a less strict dependency on `@types/webpack` like so:

```json
{
"dependencies": {
    "@types/webpack": "~4"
}
}
```

**Comment by jantimon** [comment]
I like the idea of `"@types/webpack": "~4"`

Would it also help to declare it as peer dependency?

**Comment by AviVahl** [comment]
you don't want pure-javascript users to get a peerDependency warning, so probably not. maybe a `peerDependency` with `peerDependencyMeta` as well (supported by current `npm`/`yarn` versions)?

see: https://github.com/yarnpkg/rfcs/blob/master/accepted/0000-optional-peer-dependencies.md

**Comment by jantimon** [comment]
@argv-minus-one @AviVahl do you see any possible issues for merging #1402?

**Comment by AviVahl** [comment]
yeah. I fail to see how the `~` will "fix" any duplicates for package consumers. It may even cause duplicates. That request is also confusing... Do I get the latest 4.x.x? latest 4.0.x (I believe this one is the case)?

It's important to note each `semver` request is being resolved *on its own* and locked.

So say I have a project using `@types/webpack@^4.40.0`, which was locked in my lockfile to whatever was available at the time npm resolved `"^4.40.0"`. Now when upgrading `html-webpack-plugin`, npm will now see a **new** request "~4", and will resolve that separately.

**Comment by AviVahl** [comment]
@jantimon

TBH, thinking a bit more about this, I actually don't think `html-webpack-plugin` should to install `@types/webpack` for consumers at all.

JavaScript-based projects don't really need/use them.

TypeScript-based projects should already have them installed (with own specific version) *if* they are importing `webpack` from any `.ts` file, which is the case for also importing this plugin package and using its types. Many projects don't even do that and just use untyped `webpack.config.js`.

I think `@types/webpack` should be a `"devDependency"`, leaving the versioning of it to user projects if they they-check it. It makes sense, as webpack is a `"peerDependency"`.

**Comment by criles25** [comment]
For people getting ^ in their project, I suggest setting up https://github.com/TypeStrong/fork-ts-checker-webpack-plugin. It will fix this error and significantly improve compile times.

**Comment by jonhaddow** [comment]
I noticed some plugins using an old version of Webpack. Enforcing the correct type definition version helped resolve this for me.

```
"resolutions": {
    "@types/webpack": "^4.41.17"
},
```

**Comment by Trackhe** [comment]
that problem is still alive..

```
/usr/local/bin/node /usr/local/Cellar/yarn/1.22.10/libexec/bin/yarn.js run start
yarn run v1.22.10
$ webpack serve --config webpack.dev.config.ts
[webpack-cli] Failed to load 'stuff/webpack.dev.config.ts' config
[webpack-cli] webpack.dev.config.ts:35:9 - error TS2322: Type 'HtmlWebpackPlugin' is not assignable to type '((this: Compiler, compiler: Compiler) => void) | WebpackPluginInstance'.
  Type 'HtmlWebpackPlugin' is not assignable to type 'WebpackPluginInstance'.
    Types of property 'apply' are incompatible.
      Type '(compiler: import("stuff/node_modules/html-webpack-plugin/node_modules/@types/webpack/index").Compiler) => void' is not assignable to type '(compiler: import("stuff/node_modules/webpack/types").Compiler) => void'.
        Types of parameters 'compiler' and 'compiler' are incompatible.
          Type 'Compiler' is missing the following properties from type 'Compiler': _pluginCompat, _plugins, plugin, apply, and 17 more.

35         new HtmlWebpackPlugin({
           ~~~~~~~~~~~~~~~~~~~~~~~
36             template: "src/html/index.html",
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
37         }),
   ~~~~~~~~~~

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

Process finished with exit code 2

```

**Comment by corock** [comment]
Although I don't know how webpack is configuration,
try the below steps. It works for me.

1. Install the `tsconfig-path` package.
1. Create a tsconfig-for-webpack-config.json.
1. Include TS_NODE_PROJECT path to scripts. e.g.

```json
{
  "scripts": {
    "build": "cross-env TS_NODE_PROJECT=\"tsconfig-for-webpack-config.json\" webpack"
  }
}
```

Refer to [configuration languages of the webpack docs](https://webpack.js.org/configuration/configuration-languages/#typescript).

**Comment by corock** [comment]
Try the below steps. It works for me.

1. install the `tsconfig-path` package.

```bash
npm install --save-dev tsconfig-paths
```

2. Create a tsconfig-for-webpack-config.json like this:

```json
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "esModuleInterop": true
  }
}
```

3. Include TS_NODE_PROJECT path to scripts. e.g.

```json
{
  "scripts": {
    "build": "cross-env TS_NODE_PROJECT=\"tsconfig-for-webpack-config.json\" webpack"
  }
}
```

Refer to configuration languages of the webpack [docs](https://webpack.js.org/configuration/configuration-languages/#typescript).

**Comment by haikonguyen** [comment]
How I triggered this issue is different than described in this github thread. So, in my case, our `webpack` and `tsconfig` was set properly and everything was working fine.

The Main reason that cause this issue in my case is because I locally updated `nodejs` version to  `18.14.0` , which is incompatible with old `webpack` (in our case 4.3). 

So at first, I was trying to solve the issue following various solution provided by this thread, and none helped:
    - one of the solution was required to install newer version of npm packages, which cause even more errors

It took me some time, but then I thought about the whole problem, and the main thing was working correctly, until I started locally upgrading the NodeJS

**So what helped me to solve this issue:**

1. I remove the local `node_modules` folder, which was cashed with the wrong packages
2. I have downgraded my node version to the previous one that I was using on the project (v13)
3. I also run the command **`npm cache clean --force` to clean the cache**
4. Last step was to reinstall the packages using `npm i`
5. The issue was gone.

This solved this issue for me, I hope it will help others who had similar cause of triggering this error => upgrading `nodejs`.
